**탐색적 분석**

In [2]:
import pandas as pd

# 데이터 경로
data_path = './input/'

sales_train = pd.read_csv(data_path + 'sales_train.csv')
shops = pd.read_csv(data_path + 'shops.csv')
items = pd.read_csv(data_path + 'items.csv')
item_categories = pd.read_csv(data_path + 'item_categories.csv')
test = pd.read_csv(data_path + 'test.csv')
submission = pd.read_csv(data_path + 'sample_submission.csv')

In [3]:
sales_train.head()

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0  02.01.2013               0       59    22154      999.00           1.0
1  03.01.2013               0       25     2552      899.00           1.0
2  05.01.2013               0       25     2552      899.00          -1.0
3  06.01.2013               0       25     2554     1709.05           1.0
4  15.01.2013               0       25     2555     1099.00           1.0

1. 월별 판매량을 구해야 하니 '월' 구분자만 있으면 됨
2. 2013.01.02 에 판매가(당시 판매가 999루블) 가격이 고정된게 아니라 시간이 지남에 따라 바뀔수 있다.
3. 각 상점의 상품별 일일 판매량을 월별로 합친게 월간 판매량임 
4. block_num => 2013.01월 : 0 , 2015.10월 : 34

In [5]:
# 비결측값 갯수를 표시하기 위해 show_counts = True 표시
print(f"sales train 행의 갯수 : {len(sales_train)}")
sales_train.info(show_counts=True)

sales train 행의 갯수 : 2935849
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2935849 entries, 0 to 2935848
Data columns (total 6 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   date            2935849 non-null  object 
 1   date_block_num  2935849 non-null  int64  
 2   shop_id         2935849 non-null  int64  
 3   item_id         2935849 non-null  int64  
 4   item_price      2935849 non-null  float64
 5   item_cnt_day    2935849 non-null  float64
dtypes: float64(2), int64(3), object(1)
memory usage: 134.4+ MB


1. 데이터가 300만개 가까이 되어서 메모리 134MB 임 -> 메모리사용량 줄일계획
2. 시계열 데이터는 시간흐름이 중요하므로 2013 1월부터 2015 9월까지 판매내역 훈련데이터로 사용하고
   2015 10월 판매내역 검증데이터로 사용하겠음
3. 앞장의 OOF 예측 은 시계열 데이터에서 사용할시 과거와 미래가 뒤섞이기 때문에 이용할수없다 ㅠ 

In [6]:
shops.head()

shop_name  shop_id
0   !Якутск Орджоникидзе, 56 фран        0
1   !Якутск ТЦ "Центральный" фран        1
2                Адыгея ТЦ "Мега"        2
3  Балашиха ТРК "Октябрь-Киномир"        3
4        Волжский ТЦ "Волга Молл"        4

1. 러시아어 공부 시작할순 없음
2. 다른 캐글러가 공유한 아이디어 활용할예정
3. 좀 복잡해보이니 베이스라인 짜고 성능개성 한 이후에 다룰 예정
4. 결론부터 미리 말하면 첫단어는 도시 나타냄

In [7]:
shops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   shop_name  60 non-null     object
 1   shop_id    60 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.1+ KB


상점은 딱 60 개 정도 있고 , 결측값도 없고 데이터 개수가 적어서 메모리 사용량도 적다

In [9]:
items.head()

item_name  item_id  \
0          ! ВО ВЛАСТИ НАВАЖДЕНИЯ (ПЛАСТ.)         D        0   
1  !ABBYY FineReader 12 Professional Edition Full...        1   
2      ***В ЛУЧАХ СЛАВЫ   (UNV)                    D        2   
3    ***ГОЛУБАЯ ВОЛНА  (Univ)                      D        3   
4        ***КОРОБКА (СТЕКЛО)                       D        4   

   item_category_id  
0                40  
1                76  
2                40  
3                40  
4                40

item_id 를 기준으로 slaes_train 과 item 병합이 가능할거같다. 

In [10]:
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22170 entries, 0 to 22169
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   item_name         22170 non-null  object
 1   item_id           22170 non-null  int64 
 2   item_category_id  22170 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 519.7+ KB


In [11]:
item_categories.head()

item_category_name  item_category_id
0  PC - Гарнитуры/Наушники                 0
1         Аксессуары - PS2                 1
2         Аксессуары - PS3                 2
3         Аксессуары - PS4                 3
4         Аксессуары - PSP                 4

1. category ID 기준으로 병합 가능할거같음
2. 상품분류명의 첫단어는 대분류를 뜻함 
3. 베이스 라인 끝내고 추후 피처 엔지니어링시 대분류 피처로 만들수있을듯 

병합 계획
1. sales_train 과 shops -> shop_id 기준으로 병합
2. sales_train 과 items -> item_id 기준으로 병합
3. sales_train 과 item_categories  -> item_category_id 기준으로 병합 (2번 이후에 itemcategory id 생김)

In [13]:
train = sales_train.merge(shops, on='shop_id', how='left')
train = train.merge(items, on='item_id', how='left')
train = train.merge(item_categories, on='item_category_id', how='left')

train.head().T

0  \
date                            02.01.2013   
date_block_num                           0   
shop_id                                 59   
item_id                              22154   
item_price                           999.0   
item_cnt_day                           1.0   
shop_name           Ярославль ТЦ "Альтаир"   
item_name                ЯВЛЕНИЕ 2012 (BD)   
item_category_id                        37   
item_category_name          Кино - Blu-Ray   

                                                           1  \
date                                              03.01.2013   
date_block_num                                             0   
shop_id                                                   25   
item_id                                                 2552   
item_price                                             899.0   
item_cnt_day                                             1.0   
shop_name                                Москва ТРК "Атриум"   
item_name           DEEP PURPLE  The House Of Blue Light  LP   
item_category_id                                          58   
item_category_name                            Музыка - Винил   

                                                           2  \
date                                              05.01.2013   
date_block_num                                             0   
shop_id                                                   25   
item_id                                                 2552   
item_price                                             899.0   
item_cnt_day                                            -1.0   
shop_name                                Москва ТРК "Атриум"   
item_name           DEEP PURPLE  The House Of Blue Light  LP   
item_category_id                                          58   
item_category_name                            Музыка - Винил   

                                                           3  \
date                                              06.01.2013   
date_block_num                                             0   
shop_id                                                   25   
item_id                                                 2554   
item_price                                           1709.05   
item_cnt_day                                             1.0   
shop_name                                Москва ТРК "Атриум"   
item_name           DEEP PURPLE  Who Do You Think We Are  LP   
item_category_id                                          58   
item_category_name                            Музыка - Винил   

                                                          4  
date                                             15.01.2013  
date_block_num                                            0  
shop_id                                                  25  
item_id                                                2555  
item_price                                           1099.0  
item_cnt_day                                            1.0  
shop_name                               Москва ТРК "Атриум"  
item_name           DEEP PURPLE 30 Very Best Of 2CD (Фирм.)  
item_category_id                                         56  
item_category_name      Музыка - CD фирменного производства

In [19]:
def resumetable(df):
    print(f'데이터셋 형상 : {df.shape}')
    summary = pd.DataFrame(df.dtypes, columns=['데이터 타입'])
    summary = summary.reset_index()
    summary = summary.rename(columns={'index' : '피처'})
    summary['결측값 개수'] = df.isnull().sum().values
    summary['고유값 개수'] = df.nunique().values
    summary['첫 번째 값'] = df.loc[0].values
    summary['두 번째 값'] = df.loc[1].values
    
    return summary

resumetable(train)
    

데이터셋 형상 : (2935849, 10)


피처   데이터 타입  결측값 개수  고유값 개수                  첫 번째 값  \
0                date   object       0    1034              02.01.2013   
1      date_block_num    int64       0      34                       0   
2             shop_id    int64       0      60                      59   
3             item_id    int64       0   21807                   22154   
4          item_price  float64       0   19993                   999.0   
5        item_cnt_day  float64       0     198                     1.0   
6           shop_name   object       0      60  Ярославль ТЦ "Альтаир"   
7           item_name   object       0   21807       ЯВЛЕНИЕ 2012 (BD)   
8    item_category_id    int64       0      84                      37   
9  item_category_name   object       0      84          Кино - Blu-Ray   

                                     두 번째 값  
0                                03.01.2013  
1                                         0  
2                                        25  
3                                      2552  
4                                     899.0  
5                                       1.0  
6                       Москва ТРК "Атриум"  
7  DEEP PURPLE  The House Of Blue Light  LP  
8                                        58  
9                            Музыка - Винил